eclass/docs.eclass

```diff
@@ -0,0 +1,365 @@
+# call it later in your ebuild again if necessary.
+
+# @ECLASS-VARIABLE: AUTODOC
+# @PRE_INHERIT
+# @DESCRIPTION:
+# Sets whether to use sphinx.ext.autodoc/mkautodoc
+# Defaults to 1 (True) for sphinx, and 0 (False) for mkdocs
+
+# @ECLASS-VARIABLE: OUTDIR
+# @DESCRIPTION:
+# Sets where the compiled files will be put.
+# There's no real reason to change this, but this
+# variable is useful if you want to overwrite the HTML_DOCS
+# added by this eclass. E.g.:
+#
+# HTML_DOCS=( "${yourdocs}" "${OUTDIR}/." )
+#
+# Defaults to ${DOCDIR}/_build/html
+
+# @ECLASS-VARIABLE: DOCS_CONFIG_NAME
+# @DESCRIPTION:
+# Name of the doc builder config file.
+#
+# Only relevant for doxygen, as it allows
+# config files with non-standard names
+#
+# Defaults to Doxyfile for doxygen
+
+if [[ ! ${_DOCS} ]]; then
+
+# For the python based DOCBUILDERS we need to inherit python-any-r1
+case "${DOCBUILDER}" in
+	"sphinx"|"mkdocs")
+		# If this is not a python package then
+		# this is not already set, so we need
+		# to set this to inherit python-any-r1
+		if [[ -z "${PYTHON_COMPAT}" ]]; then
+			PYTHON_COMPAT=( python3_{6,7,8} )
+		fi
+
+		# Inherit python-any-r1 if neither python-any-r1 nor
+		# python-r1 have been inherited, because we need the
+		# python_gen_any_dep function
+		if [[ ! ${_PYTHON_R1} && ! ${_PYTHON_ANY_R1} ]]; then
+			inherit python-any-r1
+		fi
+		;;
+	"doxygen")
+		# do not need to inherit anything for doxygen
+		true
+		;;
+	"")
+		die "DOCBUILDER unset, should be set to use ${ECLASS}"
+		;;
+	*)
+		die "Unsupported DOCBUILDER=${DOCBUILDER} (unknown) for ${ECLASS}"
+		;;
+esac
+
+# @FUNCTION: python_check_deps
+# @DESCRIPTION:
+# Check if the dependencies are valid
+python_check_deps() {
+	debug-print-function ${FUNCNAME}
+	use doc || return 0
+
+	local dep
+	for dep in ${check_deps[@]}; do
+		has_version "${dep}[${PYTHON_USEDEP}]" || return 1
+	done
+}
+# Save this before we start manipulating it
+check_deps=${DOCDEPEND}
+
+# @FUNCTION: python_append_dep
+# @DESCRIPTION:
+# Appends [\${PYTHON_USEDEP}] to all dependencies
+# for python based DOCBUILDERs such as mkdocs or
+# sphinx.
+python_append_deps() {
+	debug-print-function ${FUNCNAME}
+
+	local temp=()
+	local dep
+	for dep in ${DOCDEPEND[@]}; do
+		temp+=" ${dep}[\${PYTHON_USEDEP}]"
+	done
+	DOCDEPEND=${temp}
+}
+
+# @FUNCTION: sphinx_setup
+# @DESCRIPTION:
+# Sets dependencies for sphinx
+sphinx_setup() {
+	debug-print-function ${FUNCNAME}
+
+	: ${AUTODOC:=1}
+
+	if [[ ${AUTODOC} == 0 && -n "${DOCDEPEND}" ]]; then
+		die "${FUNCNAME}: do not set autodoc to 0 if external plugins are used"
+	fi
+	if [[ ${AUTODOC} == 1 ]]; then
+		DOCDEPEND="$(python_gen_any_dep "
+			dev-python/sphinx[\${PYTHON_USEDEP}]
+			${DOCDEPEND}")"
+
+	else
+		DOCDEPEND="dev-python/sphinx"
+	fi
+}
+
+# @FUNCTION: sphinx_compile
+# @DESCRIPTION:
+# Calls sphinx to build docs.
+#
+# If you overwrite src_compile or python_compile_all
+# do not call this function, call docs_compile instead
+sphinx_compile() {
```

Comment:
This is very close to collide with python eclasses. Also, I do not exactly enjoy the fact that you're duplicating functionality.

Reply:
The aim is not to duplicate the functionality of `distutils_enable_sphinx`, but to extend it to be used in ebuilds for packages that are not using distutils. I suppose I could replace the last part of this function by calling `build_sphinx "${DOCDIR}"` from the python-utils-r1 eclass since we are inheriting that anyway. But that would mean dropping the functionality to build the docs into another `OUTDIR` as the `build_sphinx` function from  python-utils-r1 does not support this. Not sure why you would want to manually specify a `OUTDIR` other than the default, but I figured I'd allow that anyway as it might prove useful in some ebuild, and because why not.

What other name would you suggest? I specifically chose `sphinx_compile` because python-utils-r1 already has `build_sphinx` and distutils-r1 has `sphinx_compile_all`. I suppose I could do `sphinx_build` but IMO that is confusing with `build_sphinx`.

Here are some ebuilds that do not inherit distutils-r1 but do have sphinx based documentation where this eclass thus came in useful:
https://gitweb.gentoo.org/repo/proj/guru.git/tree/dev-embedded/rauc/rauc-1.4.ebuild
https://gitweb.gentoo.org/repo/proj/guru.git/tree/net-vpn/riseup-vpn/riseup-vpn-0.20.4.ebuild